source/extensions/filters/network/thrift_proxy/filter.cc

```diff
@@ -0,0 +1,307 @@
+#include "extensions/filters/network/thrift_proxy/filter.h"
+
+#include "envoy/common/exception.h"
+
+#include "common/common/assert.h"
+
+#include "extensions/filters/network/thrift_proxy/buffer_helper.h"
+#include "extensions/filters/network/thrift_proxy/protocol.h"
+#include "extensions/filters/network/thrift_proxy/transport.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace NetworkFilters {
+namespace ThriftProxy {
+
+Filter::Filter(const std::string& stat_prefix, Stats::Scope& scope)
+    : req_callbacks_(*this), resp_callbacks_(*this), stats_(generateStats(stat_prefix, scope)) {}
+
+Filter::~Filter() {}
+
+void Filter::onEvent(Network::ConnectionEvent event) {
+  bool any_active_request = !active_call_list_.empty() || req_ != nullptr || resp_ != nullptr;
+
+  if (event == Network::ConnectionEvent::RemoteClose && any_active_request) {
+    stats_.cx_destroy_local_with_active_rq_.inc();
+  }
+
+  if (event == Network::ConnectionEvent::LocalClose && any_active_request) {
+    stats_.cx_destroy_remote_with_active_rq_.inc();
+  }
+}
+
+Network::FilterStatus Filter::onData(Buffer::Instance& data, bool) {
+  if (!sniffing_) {
+    if (req_buffer_.length() > 0) {
+      // Stopped sniffing during response (in onWrite). Make sure leftover req_buffer_ contents are
```

Comment:
I'm confused by the sniffing semantics. It looks like all the cases where `sniffing_` becomes false are exceptions. Is there a non-error case where it is set to false?

Reply:
I cribbed this from the mongo_proxy filter, but the idea is that for this filter (which only inspects requests/responses as they go by) if there's a protocol error we get out of the way and stop sniffing. The reasoning is that the bug may be in the Envoy decoder code falsely producing an exception on a legitimate message. I expect this to go away as we transition to a protocol-aware proxy (which will need to have any such decoder bugs sorted out).